### ISSUE TITLE
Binary size too big with s3 packages

### ISSUE THREAD

**Issue by ncw** [aws/aws-sdk-go-v2#2930]
### Acknowledgements

- [X] I have searched (https://github.com/aws/aws-sdk/issues?q=is%3Aissue) for past instances of this issue
- [X] I have verified all of my SDK modules are up-to-date (you can perform a bulk update with `go get -u github.com/aws/aws-sdk-go-v2/...`)

### Describe the bug

When I add the s3 backend which uses the SDKv2 to [rclone](https://github.com/rclone/rclone) it increases the binary by  7.4 MiB and users have been complaining about the large binary sizes.

This makes the s3 backend by far the largest contributor to the rclone binary size (about 13%). Here are the top 11 backends ranked by size measured by commenting the backend out and compiling in release mode to check the binary sizes.

Backend | Size MiB
-- | --
s3 | 7.43
storj | 4.24
protondrive | 2.61
hdfs | 2.41
oracleobjectstorage | 1.60
drive | 1.53
filescom | 1.46
azureblob | 1.26
dropbox | 1.15
googlecloudstorage | 0.78
azurefiles | 0.65

I compared this to the last version of rclone which used the SDKv1 and the s3 backend takes 6.9 MiB so this has got slightly worse. I expected this to get better with the modularization of the SDK :-(

Rclone uses the following sdk imports

```
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/aws/signer/v4"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/aws/smithy-go"
	"github.com/aws/smithy-go/logging"
	"github.com/aws/smithy-go/middleware"
	"github.com/aws/smithy-go/transport/http"
```

According to the rather neat tool [go-size-analyzer](https://github.com/Zxilly/go-size-analyzer)

The bulk of this in the service modules

![image](https://github.com/user-attachments/assets/932d0642-0e4d-4d17-8b80-4ec446adceaa)

Is there any way this can be improved?

Thank you

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Expected Behavior

I expected the S3 packages from the SDK to add no more than 1MB to my binary

### Current Behavior

It adds 7.5 MB to my binary

### Reproduction Steps

The [test_backend_sizes.py](https://github.com/rclone/rclone/blob/master/bin/test_backend_sizes.py) script will compile rclone many times to measure backend sizes, and the [go-size-analyzer](https://github.com/Zxilly/go-size-analyzer) tool can be used to verify it by running it on the rclone binary.

### AWS Go SDK V2 Module Versions Used

`go mod graph` as attachment as it is quite big.

[graph.txt](https://github.com/user-attachments/files/18115628/graph.txt)

### Compiler and Version used

go version go1.23.4 linux/amd64

### Operating System and version

Ubuntu 22.04.5 LTS

**Comment by RanVaknin** [comment]
Hi @ncw,

Thanks for reaching out. S3 is indeed a large client because it is one of the biggest AWS services in terms of its API representation size. 
We _can_ do more about trying to reduce the binary size but it will likely happen as a "side effect" of other efforts, for example https://github.com/aws/aws-sdk-go-v2/issues/2933. 

I'm going to keep this is a backlog item to gauge community interest.

Thanks,
Ran~

**Comment by jjti** [comment]
Hey just +1'ing this. Came here after trying to reduce a binary's size, found that s3 is the biggest dependency by a bit:
<img width="850" alt="Screenshot 2025-01-08 at 8 05 17 PM" src="https://github.com/user-attachments/assets/45ff4b8d-50c9-47d5-b464-03dbf8b4c8c0" />

I've got a gRPC server with OTLP and a bunch of other stuff, I was expecting some other library to come out on top, but the s3 package in particular is #1 by a fair bit at ~12% of total size

**Comment by ldez** [comment]
The problem is with all services: S3 (4.1 MB), Route53 (2.7 MB), Lighsail (4.8 MB), etc.

An important root cause of the problem is the `Client` structure: the `Client` structure has methods, and each of those methods depends on other structures, related to operations.

For the integrity of what a structure is, the compiler is forced to keep methods, and so the related structures, methods, and functions.

A way to drastically reduce the size is to remove the `Client` methods related to typed operations and replace them with functions that take the lightweight client as a parameter.

I understand that this is not a beautiful solution, and if you are using all the methods, it results in the same size; however, this is the "easiest" way, and I think nobody uses all the hundreds of methods for a client.

---

As the code is generated, maybe it's possible to hack it and apply this pattern.
I don't know how exactly [smithy](https://github.com/aws/smithy-go) works, I need to investigate, but I can see a possibility to create automatic forks with lightweight clients.

I have the feeling that modifying this block can be a first step:

https://github.com/aws/smithy-go/blob/9aa9f7326a20363bbdf6c0cc70758f5b72136e50/codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/OperationGenerator.java#L107-L120

from this:
```java
writer.openBlock("func (c $P) $T(ctx $T, params $P, optFns ...func(*Options)) ($P, error) {", "}",
...
writer.write("result, metadata, err := c.invokeOperation(ctx, $S, params, optFns, c.$L)",
```

to this:
```java
writer.openBlock("func $T(ctx $T, c $P params $P, optFns ...func(*Options)) ($P, error) {", "}",
...
writer.write("result, metadata, err := c.invokeOperation(ctx, $S, params, optFns, $L)",
```

It's possible to think the problem differently by hiding the need of the client by creating operations as structures, instead of methods or functions, with a method `invoke(c *Client)`.
But this requires a deeper change of the templates used to generate the service code.

**Comment by lucix-aws** [comment]
That would be a colossal API break so we can't do that really.

**Comment by ldez** [comment]
I never say the opposite; that's why I'm talking about creating a fork.
I was sure that creating a v3 was not a solution for you.

But my explanation of the problem is still right.

**Comment by ldez** [comment]
I tried to apply my hack.

Technically, it works, but the problem is the interaction with the elements that use interfaces based on methods of the client.
This is a problem when we need to interact with those elements like `AssumeRoleProvider`.

So, this is not a viable option for a low-maintenance fork.

I thought of ways to be non-breaking that can be applied inside this repository (no fork), and I found 2 possible solutions.
There may be other ways, but those 2 approaches seem already viable.

I don't have the full overview of the API clients, so some elements could require adjustments.

My preference is for the "lightweight client" because it produces a more significant reduction in the binary size and requires fewer changes within the current code.

---

## Solution 1: lightweight client

### Neutralize the dependencies on `Client`

1. Use an interface to abstract `Client`
    ```go
    type BaseClient interface {
        TimeOffset() *atomic.Int64
        Options() Options
    }
    ```
2. Add method `TimeOffset()` to `Client`
3. Replace usage of `*Client` with the interface
    - `finalizeOperationRetryMaxAttempts`
    - `addTimeOffsetBuild`
    - `initializeTimeOffsetResolver` -> remove it and initialize the timeout inside the constructor `NewClient`
4. Modify `stackFn` signature to `func(BaseClient, *middleware.Stack, Options)`
5. Apply new `stackFn` signature where it's necessary
    - Convert `addOperationXXX` unexposed methods into functions that use `c BaseClient` as an argument.

### Create a new lightweight client

1. Create a lightweight client `RawClient` with no "typed" methods (operation and related methods), only 3 methods:
    - `TimeOffset()`
    - `Options()`
    - `InvokeOperation()`
2. Add constructor(s) (`NewFromConfig`, `New`, etc.)
3. Expose `invokeOperation` from `RawClient`
4. Expose `addOperationXXX` functions

### Notes

To use it, the API client users have to create their method implementations based on the different exposed elements and interfaces.

This is very effective in terms of binary size reduction and has minimal changes to the existing code.

## Solution 2: Lazy Methods

### Neutralize the dependencies on `Client`

1. Use an interface to abstract `Client`
    ```go
    type BaseClient interface {
        TimeOffset() *atomic.Int64
        Options() Options
    }
    ```
2. Add method `TimeOffset()` to `Client`
3. Replace usage of `*Client` with the interface
    - `finalizeOperationRetryMaxAttempts`
    - `addTimeOffsetBuild`
    - `initializeTimeOffsetResolver` -> remove it and initialize the timeout inside the constructor `NewClient`
4. Modify `stackFn` signature to `func(BaseClient, *middleware.Stack, Options)`
5. Apply new `stackFn` signature where it's necessary
    - Convert `addOperationXXX` unexposed methods into functions that use `c BaseClient` as an argument.

### Lazy

1. Create a "factory store" inside `Client` to register `addXXX` functions
    ```go
    type Client struct {
        // ...
        factory sync.Map
      // ...
    }
    ```
2. Create one register function (not method) per exposed method.
    ```go
    func RegisterAddXXX(c *Client) {
        c.factory.Store("AddXXX", addOperationXXX)
    }
    ```
3. Use the "factory store" inside exposed methods.

    ```go
    func (c *Client) ListXXX(ctx context.Context, params *ListXXXInput, optFns ...func(*Options)) (*ListXXXOutput, error) {
        rawFunc, ok := c.factory.Load("ListXXX")
        if !ok {
            return nil, errors.New("operation ListXXX not found")
        }
    
        fn, ok := rawFunc.(func(ctx context.Context, c BaseClient, params *ListXXXInput, optFns ...func(*Options)) (*ListXXXOutput, error))
        if !ok {
            return nil, errors.New("invalid type for ListXXX operation")
        }
    
        return fn(ctx, c, params, optFns...)
    }
    ```

4. The existing interfaces related to `Client` should embed `BaseClient` interface if necessary
5. Create a "lazy" client constructor with an empty map
6. Inside the "old" `NewClient` set up the "factory" with all functions.

### Notes

This solution is less effective in binary size reduction, because the structure `ListXXXInput` and `ListXXXOutput` are still attached to the client, but the other structures (serialization/deserialization) are loaded "on-demand".
It feels less user-friendly because a user of the new constructors needs to "initialize" the "factory".

**Comment by lucix-aws** [comment]
> An important root cause of the problem is the Client structure: the Client structure has methods, and each of those methods depends on other structures, related to operations.
>
> For the integrity of what a structure is, the compiler is forced to keep methods, and so the related structures, methods, and functions.

@ldez -- are we sure that this is the case? I would expect the compiler's dead code elimination (DCE) to be able to handle unused receiver methods on structures, and some basic testing appears to confirm.

I have the following sample program that calls one SQS operation (ListQueues):

```go
package main

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
)

func main() {
	cfg, err := config.LoadDefaultConfig(context.Background())
	if err != nil {
		panic(err)
	}

	svc := sqs.NewFromConfig(cfg, func(o *sqs.Options) {
		o.ClientLogMode = aws.LogRequestWithBody | aws.LogResponseWithBody
	})

	out, err := svc.ListQueues(context.Background(), &sqs.ListQueuesInput{})
	if err != nil {
		panic(err)
	}

	fmt.Println("urls:")
	for _, q := range out.QueueUrls {
		fmt.Println(q)
	}
}
```

When I compile this program and run [go-size-analyzer](https://github.com/Zxilly/go-size-analyzer) in terminal mode, ListQueues is the only SQS operation I see included in the binary:

<img width="1613" height="256" alt="Image" src="https://github.com/user-attachments/assets/f4daab42-62e4-4a54-b5fc-03480e4354f5" />

Then, if I add a call to another operation, SendMessage (I just add a blank invocation `svc.SendMessage(context.Background(), nil)` at the end of the program), the compiled output adds the new receiver method and _only_ that new receiver method:

<img width="1599" height="314" alt="Image" src="https://github.com/user-attachments/assets/a24d81cc-032d-4746-a044-e315c5f267a3" />

In summary, **my testing appears to disprove the claim that compiler DCE cannot elide unused service operations.**

**Comment by ldez** [comment]
My previous conclusions were based on what I can see, I tried to provide a "logical" reason, maybe I was wrong.

Try by yourself with https://github.com/go-acme/lego.

- `make build` to compile the binary
- use `gsa`: you will see that **all** the `s3` operations are compiled.
- use `providers/http/s3` as example: remove the package
- edit `cmd/setup_challenges.go` to remove the case `case cmd.IsSet(flgHTTPS3Bucket):`
- compile
- use `gsa`: you will see that **all** the `s3` operations are removed.

Note that all `sts` operations are also compiled.

---

Create snippet with only the code of the package:

<details>

```go
package main

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"log"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func main() {
	c, _ := NewHTTPProvider("x")
	err := c.Present(context.Background(), "example.com", "token", "keyauth")

	if err != nil {
		log.Fatal(err)
	}
}

// HTTPProvider implements ChallengeProvider for `http-01` challenge.
type HTTPProvider struct {
	bucket string
	client *s3.Client
}

// NewHTTPProvider returns a HTTPProvider instance with a configured s3 bucket and aws session.
// Credentials must be passed in the environment variables.
func NewHTTPProvider(bucket string) (*HTTPProvider, error) {
	if bucket == "" {
		return nil, errors.New("s3: bucket name missing")
	}

	ctx := context.Background()

	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		return nil, fmt.Errorf("s3: unable to create AWS config: %w", err)
	}

	client := s3.NewFromConfig(cfg)

	return &HTTPProvider{
		bucket: bucket,
		client: client,
	}, nil
}

// Present makes the token available at `HTTP01ChallengePath(token)` by creating a file in the given s3 bucket.
func (s *HTTPProvider) Present(ctx context.Context, domain, token, keyAuth string) error {
	params := &s3.PutObjectInput{
		ACL:    "public-read",
		Bucket: aws.String(s.bucket),
		Key:    aws.String("foo"),
		Body:   bytes.NewReader([]byte(keyAuth)),
	}

	_, err := s.client.PutObject(ctx, params)
	if err != nil {
		return fmt.Errorf("s3: failed to upload token to s3: %w", err)
	}

	return nil
}

// CleanUp removes the file created for the challenge.
func (s *HTTPProvider) CleanUp(ctx context.Context, domain, token, keyAuth string) error {
	params := &s3.DeleteObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String("/foo"),
	}

	_, err := s.client.DeleteObject(ctx, params)
	if err != nil {
		return fmt.Errorf("s3: could not remove file in s3 bucket after HTTP challenge: %w", err)
	}

	return nil
}
```

</details>

- compile the binary
- use `gsa`: you will see that **only 2** of the `s3` operations are compiled.

Note even in this restricted case, the module `aws` takes 2.1 MB.

<details>

<img width="1748" height="879" alt="Image" src="https://github.com/user-attachments/assets/4b02bfef-7f4f-4c08-99e4-f568737107b5" />

</details>

---

I also tried to removed all the dependencies to `github.com/aws/aws-sdk-go-v2` (route53, lightsail) and just kept the `s3` package. The result is the same **all** the s3 operations are compiled.

<details>

<img width="1723" height="1855" alt="Image" src="https://github.com/user-attachments/assets/bbe09989-376c-451d-98a5-f4ac3a24acd8" />

</details>

---

You can do the same tests with `route53`, and `lightsail`.

---

We have all the same problem, but we are also using the same tool to analyze the sizes.

Maybe there is problem with the tool, but it all cases, if you only take a look at the compiled binary size on the FS (without gsa), each `aws` modules is taking several MB.

**Comment by lucix-aws** [comment]
I am mostly trying to make sure that we as the SDK are not _exhaustively_ defeating DCE somehow which does not appear to be the case per my SQS test. I see what you are seeing in this `go-acme/lego` program.

AFAIK if the compiler cannot _prove_ a method is not called it will not elide it. Naturally in a simple program like mine that's probably very easy for the compiler to do. In a larger program that may not be the case.

My general understanding is that uses of reflection (direct or otherwise) are the most common blocker to DCE. I do not have sufficient knowledge of this particular program to explain why DCE does not apply to the S3 client here, but I would guess that there is something in the application's calling code that is preventing it.

In the more general sense, we are pursuing artifact size reductions in 2026. As of this moment I am looking at a total serde rewrite that eliminates all of the static model-ignorant serde code and replacing it with something that can primarily live in the shared runtime.

**Comment by MartinodF** [comment]
Just to chime in, we have the same problem with EC2: we import `service/ec2`, use it as follows:

```go
client := ec2.NewFromConfig(config)
filter := ec2.DescribeInstancesInput{
	Filters: []ec2types.Filter{ /* ... running instances with a specific tag ... */ }
}

paginator := ec2.NewDescribeInstancesPaginator(client, &filter)
for paginator.HasMorePages() {
	// read instance.Tags for each instance
}
```

And that results in a 27.51MB import just for `service/ec2`, after compiling with `ldflags="-s -w"` to strip the binary. If I create a minimal reproduction DCE works well, but there's no obvious reason (e.g. `reflect` usage in the same module importing the SDK) that would explain this behavior. I'm kind of at a loss and I would appreciate AWS looking into this a bit more to understand why the SDK seems to quite often defeat DCE completely.

**Comment by ldez** [comment]
I found a tool to analyze the DCE: https://github.com/aarzilli/whydeadcode

The failure of the DCE is not related to the aws sdk itself but to other project dependencies.
A simple usage of `template/text` defeat the DCE.

The current design of the aws sdk is a problem anyway because it's very hard to remove dependencies that a project requires to avoid a failure of the DCE.
The current aws sdk relies to much on the DCE to be trimmed, IMO this should be changed.

**Comment by lucix-aws** [comment]
We have a few concrete things on the radar in 2026 that should convey substantial artifact size reductions:

1. i'm in the process of completely rewriting serialization. basically all of the statically-generated stuff in serializers.go and deserializers.go goes away, and gets replaced with a lighter set of "schemas" from the model that tell us how to do serde and delegate to protocol implementations that will live in the runtime
2. #3321

(1) is tracked over in https://github.com/aws/smithy-go/issues/458. it is a huge effort but the artifact size benefits are tangible. In a simple test program that compiles against sqs and explicitly calls every operation, `gsa` shows a reduction from ~770kb in service/sqs to ~510 as of the latest PR. I checked against ec2 as well at one point and the ratio was similar though I don't remember the exact figure. those figures are naturally subject to change as I make progress there. I also think there's a decent amount of reduction we can do on top of those two things, but that's what we're looking at in the immediate future.
